### DISCUSSION TITLE
How to custom for list view and add method DELETE in ListCreateAPIView

### DISCUSSION THREAD

**Discussion by kev26** [encode/django-rest-framework#8507] [Question & Answer] [answered]
This is my code:
```
class TutorialList(generics.ListCreateAPIView):
    queryset = Tutorial.objects.all()
    serializer_class = TutorialSerializer
    
```
I researched and found that the DELETE method in ListCreateAPIView is valid, but I don't know to use this method to delete all tutorials! And further, I want to customize the list view by filtering (title or author) instead of all tutorials!

**Comment by yyyyyyyan** [comment, marked as answer]
Hi! Let's start with the filtering, because you can also use that on DELETE!

If you want to add basic filtering via query parameters, you can override the [`get_queryset()` ](https://www.django-rest-framework.org/api-guide/generic-views/#get_querysetself) to manually add this logic:
```python
class TutorialList(generics.ListCreateAPIView):
    queryset = Tutorial.objects.all()
    serializer_class = TutorialSerializer
    filter_params = ["title", "author"]
    
    def get_queryset(self):
        queryset = super().get_queryset()
        filter_kwargs = dict()
        for param in self.filter_params:
            value = self.request.query_params.get(param)
            if value is not None:
                filter_kwargs[param] = value
        queryset = queryset.filter(**filter_kwargs)
        return queryset
```
This type of filtering, however, may be very limiting and end up being unnecessarily complex to fulfill all the required logic. If you need to use complex searches and filters, you should have a look on the DRF docs about [Generic Filtering](https://www.django-rest-framework.org/api-guide/filtering/#generic-filtering)!

Now, let's jump into the DELETE method! Since you're subclassing `ListCreateAPIView`, you already get a `.get()` method (for listing) and a `.post()` method (for creating). In order to support the HTTP DELETE method, you just need to implement a `.delete()` method in your view. You can base your code on the [`.destroy()` method of the `DestroyModelMixin`](https://github.com/encode/django-rest-framework/blob/master/rest_framework/mixins.py#L89). The only difference is that this `.destroy()` method is supposed to delete a single object from your model, so we'd need to change it a bit for it to delete a whole queryset:

```python
from rest_framework import status
from rest_framework.response import Response

class TutorialList(generics.ListCreateAPIView):
    queryset = Tutorial.objects.all()
    serializer_class = TutorialSerializer
    filter_params = ["title", "author"]
    
    def get_queryset(self):
        queryset = super().get_queryset()
        filter_kwargs = dict()
        for param in self.filter_params:
            value = self.request.query_params.get(param)
            if value is not None:
                filter_kwargs[param] = value
        queryset = queryset.filter(**filter_kwargs)
        return queryset

    def delete(self, request, *args, **kwargs):
        queryset = self.get_queryset()
        queryset.delete()
        return Response(status=status.HTTP_204_NO_CONTENT)
```

**Reply by kev26** [reply]
Thank you so much !
